### ISSUE TITLE
PIC polling does not work

### ISSUE THREAD

**Issue by Scalibq** [joncampbell123/dosbox-x#247]
I have been experimenting a bit with polling the PIC for interrupts rather than using int handlers (polling can be faster in some cases).
I have incorporated this in my SN76489 VGM player.
The int-based version works okay in DOSBox-X: https://www.dropbox.com/s/kt3acsedq2ey2v3/VGMPlay-Int.zip?dl=0
You may want to try with these Alex Kidd VGMs, as "I'm the miracle ball" and "Aaargh!" contain samples: http://www.smspower.org/Music/AlexKiddTheLostStars-SMS-PSG?sid=ac6bbd2e8efaafa46ac561197403e452
You'll need to enable the high accuracy emulation of the Tandy audio to get that to work properly.

Here is the first polled version, which I have verified to work correctly on real hardware (real 8088, real 8253 and real 8259 chips, and real SN76489AN): https://www.dropbox.com/s/q955g9mko3yfdj9/VGMPlay-polled.zip?dl=0
On DOSBox-X, it just hangs indefinitely, so apparently polling the IRR does not work like it does on real hardware.
Some background info:
I set the PIT to square wave mode (mode 3), so that its signal remains high for an extended period of time (in mode 2 it is only pulsed for 1 cycle, so you'll often miss it when you poll).
I then poll IRR until bit 1 is set, indicating the timer interrupt (interrupts are disabled, so the handler is never called):
		// Poll for interrupt
		mov dx, PIC1
		mov ah, 0x01

	pollLoop:
		in al, dx
		test al, ah
		je endPoll
// ...Unrolled for accuracy...
		in al, dx
		test al, ah
		jne pollLoop
		
	endPoll:

Then I emit the SN76489 command data.
Then I wait for the IRR to go low again, just in case (in mode 3 the new count gets effective the next time it hits 0, so I want to make sure I'm in the second half-phase of the countdown):

		// Wait for counter to go low
		mov ah, 0x01
		
	pollLoop2:
		in al, dx
		test al, ah
		je pollLoop2

And then I write a new count to the PIT for the next interrupt.

And here is the other polling version: https://www.dropbox.com/s/i1kjdp05rn9i6u7/VGMPlay-pollPIC.zip?dl=0
The code works like this:
The PIT is in rate generator mode (mode 2). I use the following code to poll the PIC for an interrupt happening (interrupts are disabled at this time, so no int handler is called, as in the other polling version):

		// Poll for interrupt
		mov dx, PIC1_COMMAND
		mov ah, 0x80
		
	pollLoop:
		// Poll for interrupt
		mov al, 0x0C//OCW3_P
		out dx, al
		in al, dx
		test al, ah
		jz pollLoop

The MSB is set when an interrupt happens, so I just test for that bit (the low bits indicate the highest int number that is currently being triggered, I just ignore that for simplicity, I only use the timer).
When you perform the poll-command, and an interrupt was signaled, the PIC treats this poll-command as if the CPU has sent an INTA to acknowledge the interrupt.
So at this point the interrupt is in service.
I have programmed the PIC to be in auto-EOI mode (in fact, the two previous versions also use auto-EOI mode), so I need to take no further action. The interrupt is automatically acknowledged and ended after the poll loop.
So I can just wait for the next interrupt in the same way.

Again, I have verified that this works correctly on real hardware. DOSBox-X hangs indefinitely, indicating that the poll-command on the PIC never sets the MSB to indicate any interrupts from the PIT.

**Comment by joncampbell123** [comment]
Thanks! So it's infinitely more complex than the simple PIC polling example I wrote in DOSLIB :)

**Comment by Scalibq** [comment]
Yes, I guess what you need is cycle-accurate emulation of the IRR and ISR states, for this kind of polling to work as expected.
The PIC emulation might execute the poll-command properly, but if the timer does not signal the interrupt flag at the right times, there is nothing to poll. I think that's what I'm seeing here (haven't checked the code yet).

**Comment by joncampbell123** [comment]
This bug reminds me... sometime later I found a PC-98 game (a sort of low resolution rail shooter) that used polling of the IRR to detect timer ticks. I was able to get it working for that game. The funny thing is the base DOSBox code that DOSBox-X was forked from mentions the possibility but says no IBM PC game used it. But, apparently the crazy devs over on the PC-98 side of things actually did! Maybe your loop works this time?
